packages/botonic-plugin-contentful/src/util/cache.ts

```diff
@@ -0,0 +1,138 @@
+import { roughSizeOfObject } from './objects'
+
+export const NOT_FOUND_IN_CACHE = Symbol('NOT_FOUND')
+
+export interface Cache<V> {
+  set(id: string, val: V): void
+  get(id: string): V | typeof NOT_FOUND_IN_CACHE
+  has(id: string): boolean
+  del(id: string): void
+
+  /**
+   * @return size in KB
+   */
+  size(): number
+
+  /**
+   * Provides the Cache keys in undefined order 1 by 1
+   * @param getMoreKeys will return true if the consumer of the keys wants another key
+   */
+  keys(): Generator<string>
+}
+
+export class InMemoryCache<V> implements Cache<V> {
+  private cache: Record<string, V> = {}
+  private sizeBytes = 0
+
+  set(id: string, val: V): void {
+    this.del(id)
+    this.sizeBytes += roughSizeOfObject(id) + roughSizeOfObject(val)
+    this.cache[id] = val
+  }
+
+  del(id: string): void {
+    const val = this.get(id)
+    if (val == NOT_FOUND_IN_CACHE) {
+      return
+    }
+    delete this.cache[id]
+    this.sizeBytes -= roughSizeOfObject(id) + roughSizeOfObject(val)
+  }
+
+  get(id: string): V | typeof NOT_FOUND_IN_CACHE {
+    // Cache.has is only checked when undefined to avoid always searching twice in the object
+    const val = this.cache[id]
+    if (val === undefined && !this.has(id)) {
+      return NOT_FOUND_IN_CACHE
+    }
+    return val
+  }
+
+  has(id: string): boolean {
+    return id in this.cache
+  }
+
+  size(): number {
+    return this.sizeBytes / 1024
+  }
+
+  len(): number {
+    return Object.keys(this.cache).length
+  }
+
+  *keys(): Generator<string> {
+    for (const k of Object.keys(this.cache)) {
+      yield k
+    }
+  }
+}
+
+/**
+ * Decorates a cache by limiting its size.
+ *
+ * TODO Use an external library to have a LRU cache. However, it's not critical
+ * because typically data in CMS is small (we're not caching media, only their
+ * URLs)
+ */
+export class LimitedCacheDecorator<V> implements Cache<V> {
+  constructor(
+    private readonly cache: Cache<V>,
+    readonly limitKB: number,
+    private readonly logger: (msg: string) => void = console.error,
+    readonly sizeWarningsRatio = [0.5, 0.75, 0.9]
+  ) {}
+
+  set(id: string, val: V): void {
+    const incBytes = roughSizeOfObject(id) + roughSizeOfObject(val)
+    const checkFit = () =>
+      this.cache.size() * 1024 + incBytes <= this.limitKB * 1024
+    let itFits = checkFit()
+
+    if (!itFits) {
+      for (const k of this.cache.keys()) {
+        if (itFits) {
+          break
+        }
+        this.cache.del(k)
+        itFits = checkFit()
+      }
+      if (!itFits) {
+        this.logger(
+          `Cannot add object in cache because it's larger than max capacity(${this.limitKB})`
+        )
+        return
+      }
```

Comment:
Will this code ever be executed? 
In the previous loop, if adding the new key doesn't fit in the cache, it will keep removing cached values until it fits, so at the very end the cache may not have any value so it will be able to add new values because it will be empty, right?

Reply:
since checkFit takes into account the incremental size of the new item, it's theoretically possible that the single item size is larger that the full cache capacity.
I improved the error message to be more descriptive
```
const checkFit = () =>
      this.cache.size() * 1024 + incBytes <= this.limitKB * 1024
    
```

Reply:
Thanks for the clarification @dpinol!